File: core/src/core/retrieval_telemetry.py
"""Retrieval decision/feedback telemetry for the Python agent surface.

Mirrors the VS Code extension's MCP telemetry (vinv-vs
``src/mcp/retrievalTelemetry.ts``) byte-for-byte where it matters, so both
surfaces write compatible events into the same ledger
(``~/.vinv/telemetry/retrieval.jsonl``) and the offline policy evaluator
(``index/eval/off_policy.py``) learns from Python-agent retrievals exactly as
it does from editor retrievals:

* the epoch is the identical sha256 over the store's ``meta.json`` +
  ``manifest.json`` (24 hex chars), so events from both surfaces join;
* query features are content-safe (hash + shape only — never the query text);
* epsilon-greedy exploration uses the same env knobs
  (``VINV_RETRIEVAL_POLICY_MODE``, ``VINV_RETRIEVAL_ACTIONS``,
  ``VINV_RETRIEVAL_EPSILON``) and logs exact per-action propensities.
"""

from __future__ import annotations

import hashlib
import json
import math
import os
import random
import re
import time
import uuid
from pathlib import Path
from typing import Any

_CODE_IDENTIFIER = re.compile(r"[_:.()\[\]{}]")


def _vinv_home() -> Path:
    override = os.environ.get("VINV_HOME", "").strip()
    return Path(override) if override else Path.home() / ".vinv"


def _telemetry_path() -> Path:
    return _vinv_home() / "telemetry" / "retrieval.jsonl"


def retrieval_epoch(store_dir: str | os.PathLike[str]) -> str:
    """Identical to the extension's epoch: sha256(meta.json + manifest.json)."""
    digest = hashlib.sha256()
    store = Path(store_dir)
    for filename in ("meta.json", "manifest.json"):
        try:
            digest.update((store / filename).read_bytes())
        except OSError:
            digest.update(f"{filename}:missing".encode())
    return digest.hexdigest()[:24]


def query_features(query: str) -> dict[str, Any]:
    """Content-safe query shape — hash and counts, never the text."""
    words = [w for w in query.strip().split() if w]
    return {
        "query_hash": hashlib.sha256(query.encode()).hexdigest()[:24],
        "char_count": len(query),
        "token_estimate": math.ceil(len(query) / 4),
        "word_count": len(words),
        "has_code_identifier": any(_CODE_IDENTIFIER.search(w) for w in words),
    }


def retrieval_action_set(requested_top_k: int) -> list[int]:
    raw = os.environ.get("VINV_RETRIEVAL_ACTIONS", "3,5,8,10")
    parsed = []
    for part in raw.split(","):
        try:
            value = int(part.strip())
        except ValueError:
            continue
        if 1 <= value <= 20:
            parsed.append(value)
    actions = parsed or [3, 5, 8, 10]
    if requested_top_k not in actions:
        actions.append(requested_top_k)
    return sorted(set(actions))


def exploration_epsilon() -> float:
    try:
        raw = float(os.environ.get("VINV_RETRIEVAL_EPSILON", "0.1"))
    except ValueError:
        return 0.1
    return raw if 0 <= raw <= 0.5 else 0.1


def select_retrieval_action(
    requested_top_k: int, random_value: float | None = None
) -> dict[str, Any]:
    """Epsilon-greedy action selection with exact logged propensities.

    The Python surface has no canary/policy file plumbing (that lives in the
    editor); it participates in learning through the explore/off modes.
    propensity(a) = eps/|A| + (1-eps)*[a == requested].
    """
    mode = os.environ.get("VINV_RETRIEVAL_POLICY_MODE", "shadow")
    if mode != "explore":
        return {"top_k": requested_top_k, "propensity": 1.0, "policy": "baseline"}
    actions = retrieval_action_set(requested_top_k)
    epsilon = exploration_epsilon()
    draw = random.random() if random_value is None else random_value
    chosen = requested_top_k
    if epsilon > 0 and draw < epsilon:
        index = min(int((draw / epsilon) * len(actions)), len(actions) - 1)
        chosen = actions[index]
    propensity = epsilon / len(actions) + (
        1 - epsilon if chosen == requested_top_k else 0.0
    )
    return {
        "top_k": chosen,
        "propensity": propensity,
        "policy": "baseline" if chosen == requested_top_k else "explore",
    }


def append_retrieval_event(event: dict[str, Any]) -> None:
    """Append one event; telemetry must never break retrieval itself."""
    target = _telemetry_path()
    try:
        target.parent.mkdir(parents=True, exist_ok=True)
        with open(target, "a", encoding="utf-8") as handle:
            handle.write(json.dumps(event) + "\n")
        os.chmod(target, 0o600)
    except OSError:
        pass


def result_fingerprints(results: list[dict[str, Any]]) -> tuple[list[str], list[str]]:
    """(hashes, files) for the first 20 hits — same shape as the extension."""
    hashes = [
        hashlib.sha256(
            json.dumps(result, separators=(",", ":"), sort_keys=False).encode()
        ).hexdigest()[:24]
        for result in results[:20]
    ]
    files: list[str] = []
    for result in results[:20]:
        file = result.get("file")
        if isinstance(file, str) and file and file not in files:
            files.append(file)
    return hashes, files


def log_retrieval_decision(
    *,
    store_dir: str | os.PathLike[str],
    query: str,
    requested_top_k: int,
    selection: dict[str, Any],
    latency_ms: float,
    results: list[dict[str, Any]],
) -> str:
    """Log one decision event; returns the decision id for feedback joins."""
    decision_id = uuid.uuid4().hex
    hashes, _files = result_fingerprints(results)
    append_retrieval_event(
        {
            "type": "decision",
            "ts": time.strftime("%Y-%m-%dT%H:%M:%S", time.gmtime())
            + f".{int(time.time() * 1000) % 1000:03d}Z",
            "decision_id": decision_id,
            "epoch": retrieval_epoch(store_dir),
            "surface": "python-agent",
            "features": query_features(query),
            "action": {"top_k": selection["top_k"], "policy": selection["policy"]},
            "served_top_k": selection["top_k"],
            "propensity": selection["propensity"],
            "fallback": False,
            "shadow_action": None,
            "latency_ms": round(latency_ms, 3),
            "result_count": len(results),
            "result_hashes": hashes,
        }
    )
    return decision_id


def record_retrieval_feedback(
    decision_id: str,
    reward: float,
    *,
    store_dir: str | os.PathLike[str],
    outcome: str | None = None,
) -> None:
    """Explicit reward for an earlier decision (same schema as vinv_feedback)."""
    append_retrieval_event(
        {
            "type": "feedback",
            "ts": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
            "decision_id": decision_id,
            "epoch": retrieval_epoch(store_dir),
            "source": "explicit",
            "reward": float(reward),
            "outcome": (outcome or "")[:80] or None,
        }
    )


__all__ = [
    "append_retrieval_event",
    "exploration_epsilon",
    "log_retrieval_decision",
    "query_features",
    "record_retrieval_feedback",
    "result_fingerprints",
    "retrieval_action_set",
    "retrieval_epoch",
    "select_retrieval_action",
]

File: extension/src/mcp/retrievalTelemetry.ts
import * as crypto from 'crypto';
import * as fs from 'fs';
import * as path from 'path';
import { vinvHomeDir } from '../vinvHome';

export interface RetrievalEvent {
	type:
		| 'decision'
		| 'feedback'
		| 'shadow'
		| 'policy_invalidated'
		| 'rollback'
		// QnA answer-quality telemetry: the deterministic critic's report per
		// attempt, and the anomaly of a missing/malformed sufficiency verdict.
		| 'critic'
		| 'verdict_anomaly'
		// One automatic off-policy evaluation over the pooled ledger: the
		// per-candidate DR deltas, gates, and what (if anything) was promoted.
		| 'ope_evaluation'
		// A thumbs-down answer's surfaced files — accumulated negative alias
		// evidence consumed by tag enrichment (never acted on singly).
		| 'negative_tag_evidence';
	ts: string;
	decision_id?: string;
	epoch: string;
	[key: string]: unknown;
}

export interface RetrievalPolicy {
	version: 1;
	epoch: string;
	top_k: number;
	canary_fraction: number;
}

export interface PolicyDecision {
	topK: number;
	propensity: number;
	policy: 'baseline' | 'canary' | 'active' | 'explore';
	shadowTopK: number | null;
}

/** A decision awaiting explicit or auxiliary (edit-overlap) feedback. */
export interface PendingDecision {
	decision_id: string;
	epoch: string;
	ts: string;
	canary: boolean;
	files: string[];
}

/** Durable MCP learning state — survives server restarts so a rolled-back
 * canary stays rolled back and pending decisions still receive feedback. */
export interface TelemetryState {
	version: 1;
	policy_disabled: boolean;
	consecutive_canary_losses: number;
	pending: PendingDecision[];
}

/**
 * The discrete action set for top-k exploration. Env-tunable via
 * VINV_RETRIEVAL_ACTIONS (comma-separated); the requested top_k is always a
 * member so the greedy arm has non-zero propensity.
 */
export function retrievalActionSet(requestedTopK: number): number[] {
	const raw = process.env.VINV_RETRIEVAL_ACTIONS ?? '3,5,8,10';
	const parsed = raw
		.split(',')
		.map((value) => Number.parseInt(value.trim(), 10))
		.filter((value) => Number.isInteger(value) && value >= 1 && value <= 20);
	const actions = parsed.length > 0 ? parsed : [3, 5, 8, 10];
	if (!actions.includes(requestedTopK)) {
		actions.push(requestedTopK);
	}
	return [...new Set(actions)].sort((a, b) => a - b);
}

/** Exploration rate for epsilon-greedy logging (VINV_RETRIEVAL_EPSILON). */
export function explorationEpsilon(): number {
	const raw = Number.parseFloat(process.env.VINV_RETRIEVAL_EPSILON ?? '0.1');
	return Number.isFinite(raw) && raw >= 0 && raw <= 0.5 ? raw : 0.1;
}

export function retrievalEpoch(storeDir: string): string {
	const hash = crypto.createHash('sha256');
	for (const filename of ['meta.json', 'manifest.json']) {
		try {
			hash.update(fs.readFileSync(path.join(storeDir, filename)));
		} catch {
			hash.update(`${filename}:missing`);
		}
	}
	return hash.digest('hex').slice(0, 24);
}

export function queryFeatures(query: string): Record<string, number | boolean | string> {
	const words = query.trim().split(/\s+/).filter(Boolean);
	return {
		query_hash: crypto.createHash('sha256').update(query).digest('hex').slice(0, 24),
		char_count: query.length,
		token_estimate: Math.ceil(query.length / 4),
		word_count: words.length,
		has_code_identifier: words.some((word) => /[_:.()[\]{}]/.test(word)),
	};
}

export function appendRetrievalEvent(event: RetrievalEvent): void {
	const directory = path.join(vinvHomeDir(), 'telemetry');
	const target = path.join(directory, 'retrieval.jsonl');
	fs.mkdirSync(directory, { recursive: true });
	const descriptor = fs.openSync(target, 'a', 0o600);
	try {
		fs.writeSync(descriptor, `${JSON.stringify(event)}\n`);
	} finally {
		fs.closeSync(descriptor);
	}
	try {
		fs.chmodSync(target, 0o600);
	} catch {
		// chmod is not meaningful on every platform.
	}
}

export function loadPolicyForEpoch(epoch: string): RetrievalPolicy | null {
	const target = path.join(vinvHomeDir(), 'retrieval-policy.json');
	try {
		const policy = JSON.parse(fs.readFileSync(target, 'utf8')) as Partial<RetrievalPolicy>;
		if (
			policy.version !== 1 ||
			// 'any' = an epoch-agnostic policy from a POOLED evaluation (the
			// top-k action's semantics are epoch-stable). The previous
			// exact-match-only rule meant every promoted policy died on the
			// next index update, making promotion structurally pointless.
			(policy.epoch !== epoch && policy.epoch !== 'any') ||
			typeof policy.top_k !== 'number' ||
			!Number.isInteger(policy.top_k) ||
			policy.top_k < 1 ||
			policy.top_k > 20 ||
			(policy.canary_fraction !== undefined &&
				(typeof policy.canary_fraction !== 'number' ||
					policy.canary_fraction <= 0 ||
					policy.canary_fraction > 0.1))
		) {
			appendRetrievalEvent({
				type: 'policy_invalidated',
				ts: new Date().toISOString(),
				epoch,
				stored_epoch: policy.epoch ?? null,
			});
			return null;
		}
		return {
			version: 1,
			epoch,
			top_k: policy.top_k,
			canary_fraction: policy.canary_fraction ?? 0.05,
		};
	} catch {
		return null;
	}
}

export function selectRetrievalAction(
	requestedTopK: number,
	policy: RetrievalPolicy | null,
	mode: string,
	randomValue: number = Math.random(),
): PolicyDecision {
	if (mode === 'explore') {
		// Epsilon-greedy over the discrete action set with exact propensities:
		// with probability 1-eps serve the requested top_k, otherwise a uniform
		// draw over the set. propensity(a) = eps/|A| + (1-eps)*[a == requested],
		// which identifies IPS/SNIPS/DR for every action in the set.
		const actions = retrievalActionSet(requestedTopK);
		const epsilon = explorationEpsilon();
		let chosen = requestedTopK;
		if (epsilon > 0 && randomValue < epsilon) {
			// Given randomValue < eps, randomValue/eps is uniform on [0,1).
			const index = Math.min(
				Math.floor((randomValue / epsilon) * actions.length),
				actions.length - 1,
			);
			chosen = actions[index];
		}
		const propensity =
			epsilon / actions.length + (chosen === requestedTopK ? 1 - epsilon : 0);
		return {
			topK: chosen,
			propensity,
			policy: chosen === requestedTopK ? 'baseline' : 'explore',
			shadowTopK: policy && policy.top_k !== chosen ? policy.top_k : null,
		};
	}
	if (!policy || mode === 'off') {
		return { topK: requestedTopK, propensity: 1, policy: 'baseline', shadowTopK: null };
	}
	if (mode === 'active') {
		return { topK: policy.top_k, propensity: 1, policy: 'active', shadowTopK: null };
	}
	if (mode === 'canary' && policy.top_k !== requestedTopK) {
		const useCanary = randomValue < policy.canary_fraction;
		return useCanary
			? {
					topK: policy.top_k,
					propensity: policy.canary_fraction,
					policy: 'canary',
					shadowTopK: null,
				}
			: {
					topK: requestedTopK,
					propensity: 1 - policy.canary_fraction,
					policy: 'baseline',
					shadowTopK: policy.top_k,
				};
	}
	return {
		topK: requestedTopK,
		propensity: 1,
		policy: 'baseline',
		shadowTopK: policy.top_k,
	};
}

const EMPTY_STATE: TelemetryState = {
	version: 1,
	policy_disabled: false,
	consecutive_canary_losses: 0,
	pending: [],
};

function stateFilePath(): string {
	return path.join(vinvHomeDir(), 'telemetry', 'state.json');
}

/** Load durable learning state; a missing or corrupt file yields a clean slate. */
export function loadTelemetryState(): TelemetryState {
	try {
		const raw = JSON.parse(fs.readFileSync(stateFilePath(), 'utf8')) as Partial<TelemetryState>;
		if (raw.version !== 1) {
			return { ...EMPTY_STATE, pending: [] };
		}
		return {
			version: 1,
			policy_disabled: raw.policy_disabled === true,
			consecutive_canary_losses:
				typeof raw.consecutive_canary_losses === 'number' &&
				Number.isInteger(raw.consecutive_canary_losses) &&
				raw.consecutive_canary_losses >= 0
					? raw.consecutive_canary_losses
					: 0,
			pending: Array.isArray(raw.pending)
				? raw.pending.filter(
						(entry): entry is PendingDecision =>
							typeof entry === 'object' &&
							entry !== null &&
							typeof (entry as PendingDecision).decision_id === 'string' &&
							typeof (entry as PendingDecision).epoch === 'string' &&
							typeof (entry as PendingDecision).ts === 'string' &&
							Array.isArray((entry as PendingDecision).files),
					)
				: [],
		};
	} catch {
		return { ...EMPTY_STATE, pending: [] };
	}
}

/** Atomically persist learning state so an MCP restart cannot resurrect a
 * rolled-back canary or orphan pending decisions. */
export function saveTelemetryState(state: TelemetryState): void {
	const target = stateFilePath();
	fs.mkdirSync(path.dirname(target), { recursive: true });
	const temporary = `${target}.${process.pid}.tmp`;
	fs.writeFileSync(temporary, JSON.stringify(state), { mode: 0o600 });
	fs.renameSync(temporary, target);
}
